### PR TITLE
Trusted publishing: use `user/repo` slug in GitHub publisher form

### LINKED ISSUE #13660: Trusted publishing: consolidate the GitHub publisher form a little
The registration forms for GitHub publishers currently look like this:

image

The split between "Owner" and "Repository name" is arguably pointless: the `user/repo` slug format is well known and commonly used by developers, and having it broken across multiple form inputs means that the user needs to tab or otherwise select between them.

So, my proposal: remove the separate form inputs here, and consolidate them into a single "Repository" input that expects inputs in `user/repo` slug format.

This would mean updating the screenshots once more, but IMO it's worth it for the slightly improved UX 🙂 

CC @di for thoughts

### PR DESCRIPTION
**Pull request by jleightcap**
Change form representation to use a singular "user/repo" format.

- Merge `GitHubPublisherBase`'s `"owner"` and `"repository"` fields into the `"repo_slug"`.
- Update UI templates.

Resolves #13660

### CODE REVIEW

**Comment by woodruffw** [comment]
> I'm no "UX/DX" developer, so I'd like to get confirmation this makes sense. The "Repository" field asking for a "user/repo(sitory)" might be seem a little recursive, but like expressed previously is well known and used.

That part looks fine to me -- let's also update the hint inside the `Repository` form input so that it says `user/repo` instead of just `repository`.

File: warehouse/accounts/views.py [hunk 1]
```diff
@@ -1462,8 +1462,7 @@ def add_pending_github_oidc_publisher(self):
         publisher_already_exists = (
             self.request.db.query(PendingGitHubPublisher)
             .filter_by(
-                repository_name=form.repository.data,
-                repository_owner=form.normalized_owner,
+                repository=form.repository.data,
```
Comment (woodruffw) [inline warehouse/accounts/views.py:1465]: This isn't going to work, unfortunately -- `repository` isn't a field on a `PendingGitHubPublisher`.

Instead, we should do something similar to what was already in place: the form model should provide `normalized_owner` and `normalized_repository` attributes that get computed from `form.repository.data`.

File: warehouse/templates/manage/account/publishing.html [hunk 2]
```diff
@@ -134,33 +134,18 @@ <h3>GitHub</h3>
...
     <div class="form-group">
       <label for="repository" class="form-group__label">
-        {% trans %}Repository name{% endtrans %}
+        {% trans %}Repository{% endtrans %}
         {% if pending_github_publisher_form.repository.flags.required %}
         <span class="form-group__required">{% trans %}(required){% endtrans %}</span>
         {% endif %}
       </label>
-      {{ pending_github_publisher_form.repository(placeholder=gettext("repository"), autocomplete="off", autocapitalize="off", spellcheck="false", class_="form-group__field", **{"aria-describedby":"repository-errors"}) }}
+      {{ pending_github_publisher_form.owner(placeholder=gettext("repository"), autocomplete="off", autocapitalize="off", spellcheck="false", class_="form-group__field", aria_describedby="owner-errors") }}
```
Comment (woodruffw) [inline warehouse/templates/manage/account/publishing.html:144]: Let's update the placeholder text to hint at the format we're expecting.

```suggestion
      {{ pending_github_publisher_form.owner(placeholder=gettext("user/repo"), autocomplete="off", autocapitalize="off", spellcheck="false", class_="form-group__field", aria_describedby="owner-errors") }}
```

**Review by woodruffw — COMMENTED** [review]
Render itself LGTM here (besides one nit) -- this changeset is also going to need changes to the form itself, which is under `warehouse/oidc/forms.py`.

File: warehouse/templates/manage/project/publishing.html [hunk 3]
```diff
@@ -71,31 +71,16 @@ <h3>GitHub</h3>
...
+          {% trans %}Repository{% endtrans %}
           {% if github_publisher_form.repository.flags.required %}
           <span class="form-group__required">{% trans %}(required){% endtrans %}</span>
           {% endif %}
         </label>
-        {{ github_publisher_form.repository(placeholder=gettext("repository"), autocomplete="off", autocapitalize="off", spellcheck="false", class_="form-group__field", **{"aria-describedby":"repository-errors"}) }}
+        {{ github_publisher_form.repository(placeholder=gettext("user/repo"), autocomplete="off", autocapitalize="off", spellcheck="false", class_="form-group__field", **{"aria-describedby":"repository-errors"}) }}
         <p class="form-group__help-text">
-          {% trans %}The name of the GitHub repository that contains the publishing workflow{% endtrans %}
+          {% trans %}The GitHub repository (user/repo) that contains the publishing workflow{% endtrans %}
```
Comment (woodruffw) [inline warehouse/templates/manage/project/publishing.html:83]:
```suggestion
          {% trans %}The GitHub repository (in user/repo format) that contains the publishing workflow{% endtrans %}
```

File: warehouse/oidc/forms.py [hunk 4]
```diff
@@ -20,26 +20,21 @@
 from warehouse.i18n import localize as _
 from warehouse.utils.project import PROJECT_NAME_RE
 
-_VALID_GITHUB_REPO = re.compile(r"^[a-zA-Z0-9-_.]+$")
-_VALID_GITHUB_OWNER = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9-]*$")
+_VALID_GITHUB_REPO = "[a-zA-Z0-9-_.]+"
+_VALID_GITHUB_OWNER = "[a-zA-Z0-9][a-zA-Z0-9-]*"
```
Comment (woodruffw) [inline warehouse/oidc/forms.py:24]: Nitpick: use r-strings here (not strictly necessary, but IDEs will render them more nicely)

```suggestion
_VALID_GITHUB_REPO = r"[a-zA-Z0-9-_.]+"
_VALID_GITHUB_OWNER = r"[a-zA-Z0-9][a-zA-Z0-9-]*"
```

File: warehouse/oidc/forms.py [hunk 5]
```diff
@@ -122,21 +117,17 @@ def _lookup_owner(self, owner):
...
+        # Compute DB's `repository_name` `repository_owner` from the form's `repo_slug`.
+        repo_slug = field.data.split("/")
+        (owner, repository) = (repo_slug[0], repo_slug[1])
 
         owner_info = self._lookup_owner(owner)
 
         # NOTE: Use the normalized owner name as provided by GitHub.
         self.normalized_owner = owner_info["login"]
         self.owner_id = owner_info["id"]
+        self.normalized_repository = repository
```
Comment (woodruffw) [inline warehouse/oidc/forms.py:130]: Nitpick: We haven't actually done any normalization here, so let's just call this `self.repository` 🙂

Comment (woodruffw) [reply warehouse/oidc/forms.py:130]: (There probably _is_ some normalization we could do here, but it isn't super well defined: GitHub URLs are _somewhat_ case sensitive, but I don't think the repository name in the claim is normalized in any way.)

File: warehouse/oidc/forms.py [hunk 6]
```diff
@@ -122,21 +114,22 @@ def _lookup_owner(self, owner):
...
-        owner = field.data
-
-        # We pre-filter owners with a regex, to avoid loading GitHub's API
-        # with usernames/org names that will never be valid.
-        if not _VALID_GITHUB_OWNER.match(owner):
+    def validate_repo_slug(self, field):
+        if not _VALID_GITHUB_SLUG.match(field.data):
             raise wtforms.validators.ValidationError(
-                _("Invalid GitHub user or organization name.")
+                _("Invalid repository")
```
Comment (woodruffw) [inline warehouse/oidc/forms.py:120]: Let's make this less generic:

```suggestion
                _("Invalid repository: not in user/repo format")
```

File: tests/unit/oidc/test_forms.py [hunk 7]
```diff
@@ -288,20 +286,31 @@ def test_validate(self, monkeypatch):
...
-        "workflow_filename", ["missing_suffix", "/slash", "/many/slashes", "/slash.yml"]
+        "workflow_filename",
+        [
+            "missing_suffix",
+            "/slash",
+            "/many/slashes",
+            "/slash.yml",
+            "repo@/badsym",
+            "user/@badsym",
+            "-user/...",
```
Comment (woodruffw) [inline tests/unit/oidc/test_forms.py:312]: I think you meant to put these last three testcases in `test_validate_repo_slug`, not here 🙂

Comment (woodruffw) [reply tests/unit/oidc/test_forms.py:312]: (I can move them.)

Comment (woodruffw) [reply tests/unit/oidc/test_forms.py]: 836aa9dac..f3c4090dd

**Review by woodruffw — APPROVED** [review]
LGTM, nice work @jleightcap!

**Review by dstufft — APPROVED** [review]

**Comment by di** [comment]
Closing per https://github.com/pypi/warehouse/issues/13660#issuecomment-1604347169.

**Pull request closed without merging**